{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json, os\n",
    "import shutil\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_tcga_data(json_file, source_dirctory, target_dirctory):\n",
    "\n",
    "    if not json_file or json_file == '':\n",
    "        return False\n",
    "    if not os.path.exists(source_dirctory):\n",
    "        return False\n",
    "    if not os.path.exists(target_dirctory):\n",
    "        os.makedirs(target_dirctory)\n",
    "\n",
    "    # read json_file\n",
    "    with open(json_file, 'r') as jf:\n",
    "        json_data = jf.read()\n",
    "\n",
    "    json_info = json.loads(json_data)\n",
    "\n",
    "    for file_info in json_info:\n",
    "        file_id = file_info['file_id']\n",
    "        file_name = file_info['file_name']\n",
    "        submit_id = file_info['associated_entities'][0]['entity_submitter_id']\n",
    "        source_file = source_dirctory + '/' + file_id + '/' + file_name\n",
    "        target_file = target_dirctory + '/' + submit_id + '.tsv'\n",
    "        shutil.copy(source_file, target_file)\n",
    "        print('file_name :{} copy done'.format(file_name))\n",
    "\n",
    "# preprocess_tcga_data(\"D:\\BTp\\data\\metadata.cart.2023-03-02.json\",\"D:\\BTp\\data\",\"D:\\BTp\\processdata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "#remove unecessary data \n",
    "def prodata(source_dirctory, target_dirctory):\n",
    "    if not os.path.exists(source_dirctory):\n",
    "        return False\n",
    "    if not os.path.exists(target_dirctory):\n",
    "        os.makedirs(target_dirctory)\n",
    "    files = os.listdir(source_dirctory)\n",
    "\n",
    "    for file in files:\n",
    "        if file.endswith('.tsv'):\n",
    "            file_path = os.path.join(source_dirctory, file)\n",
    "            df = pd.read_csv(file_path, sep='\\t', header=None, skiprows=[\n",
    "                             0, 2, 3, 4, 5], usecols=[0, 1, 2, 3])\n",
    "            f = os.path.splitext(file)[0]\n",
    "            df.to_csv(target_dirctory+'/'+f, index=False, header=False)\n",
    "            # print(f,\" done\")\n",
    "\n",
    "prodata(\"D:\\BTp\\processdata\",\"D:\\BTp\\Process1data\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Sachin Meena\\AppData\\Local\\Temp\\ipykernel_11720\\840328527.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  GFF3 = pd.read_csv(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3074189</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>AC240274.1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3074340</th>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>U1</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3074343</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>U1</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3074346</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>AC213203.2</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3074354</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>AC213203.1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60666 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene_id    gene_name                        gene_biotype\n",
       "0        ENSG00000223972      DDX11L1  transcribed_unprocessed_pseudogene\n",
       "12       ENSG00000227232       WASH7P              unprocessed_pseudogene\n",
       "25       ENSG00000278267    MIR6859-1                               miRNA\n",
       "28       ENSG00000243485  MIR1302-2HG                              lncRNA\n",
       "36       ENSG00000284332    MIR1302-2                               miRNA\n",
       "...                  ...          ...                                 ...\n",
       "3074189  ENSG00000271254   AC240274.1                      protein_coding\n",
       "3074340  ENSG00000275405           U1                               snRNA\n",
       "3074343  ENSG00000275987           U1                               snRNA\n",
       "3074346  ENSG00000277475   AC213203.2                      protein_coding\n",
       "3074354  ENSG00000268674   AC213203.1                      protein_coding\n",
       "\n",
       "[60666 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#collect all genes information from gtf file\n",
    "GFF3 = pd.read_csv(\n",
    "    filepath_or_buffer='Homo_sapiens.GRCh38.103.gtf',\n",
    "    sep='\\t',\n",
    "    header=None,\n",
    "    names=['seqid', 'source', 'type', 'start', 'end',\n",
    "           'score', 'strand', 'phase', 'attributes'],\n",
    "    skiprows=[0, 1, 2, 3, 4]\n",
    ")\n",
    "GFF3 = GFF3[GFF3['source'].notnull()]\n",
    "\n",
    "GFF3 = GFF3[GFF3['type'] == 'gene']\n",
    "\n",
    "GFF3['gene_id'] = GFF3.apply(lambda x: x.attributes.split(\n",
    "    ';')[0].strip().split(' ')[1].strip('\"'), axis=1)\n",
    "GFF3['gene_name'] = GFF3.apply(lambda x: x.attributes.split(\n",
    "    ';')[2].strip().split(' ')[1].strip('\"'), axis=1)\n",
    "GFF3['gene_biotype'] = GFF3.apply(lambda x: x.attributes.split(\n",
    "    ';')[4].strip().split(' ')[1].strip('\"'), axis=1)\n",
    "\n",
    "GFF3 = pd.DataFrame(GFF3, columns=['gene_id', 'gene_name', 'gene_biotype'])\n",
    "GFF3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "cv=0\n",
    "for file_name in os.listdir('D:\\BTp\\Process1data'):\n",
    "    if file_name.endswith('.csv'):\n",
    "        cv+=1\n",
    "    else :\n",
    "        file_path = os.path.join('D:\\BTp\\Process1data', file_name)\n",
    "        df = pd.read_csv(file_path, header=None,skiprows=1)\n",
    "        df.drop(df.columns[[1,2]], axis=1, inplace=True)\n",
    "        df.to_csv(file_path, header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA-05-4244-01A-01R-1107-07  merge completed!\n",
      "TCGA-05-4249-01A-01R-1107-07  merge completed!\n",
      "TCGA-05-4250-01A-01R-1107-07  merge completed!\n",
      "TCGA-05-4382-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4384-01A-01R-1755-07  merge completed!\n",
      "TCGA-05-4389-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4390-01A-02R-1755-07  merge completed!\n",
      "TCGA-05-4395-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4396-01A-21R-1858-07  merge completed!\n",
      "TCGA-05-4397-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4398-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4402-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4403-01A-01R-1206-07  merge completed!\n",
      "TCGA-05-4405-01A-21R-1858-07  merge completed!\n",
      "TCGA-05-4410-01A-21R-1858-07  merge completed!\n",
      "TCGA-05-4415-01A-22R-1858-07  merge completed!\n",
      "TCGA-05-4417-01A-22R-1858-07  merge completed!\n",
      "TCGA-05-4418-01A-01R-1206-07  merge completed!\n"
     ]
    },
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 353. MiB for an array with shape (2, 23129217) and data type object",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 15\u001b[0m\n\u001b[0;32m     11\u001b[0m         \u001b[39mprint\u001b[39m(file,\u001b[39m'\u001b[39m\u001b[39m merge completed!\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m     12\u001b[0m     \u001b[39mreturn\u001b[39;00m df\n\u001b[1;32m---> 15\u001b[0m dff \u001b[39m=\u001b[39m all_data(\u001b[39m\"\u001b[39;49m\u001b[39mD:\u001b[39;49m\u001b[39m\\\u001b[39;49m\u001b[39mBTp\u001b[39;49m\u001b[39m\\\u001b[39;49m\u001b[39mProcess1data\u001b[39;49m\u001b[39m\"\u001b[39;49m, GFF3)\n",
      "Cell \u001b[1;32mIn[6], line 10\u001b[0m, in \u001b[0;36mall_data\u001b[1;34m(source_dirctory, df)\u001b[0m\n\u001b[0;32m      8\u001b[0m     data \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(filepath_or_buffer\u001b[39m=\u001b[39msource_dirctory \u001b[39m+\u001b[39m \u001b[39m'\u001b[39m\u001b[39m/\u001b[39m\u001b[39m'\u001b[39m \u001b[39m+\u001b[39m file,sep\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m,\u001b[39m\u001b[39m'\u001b[39m, names\u001b[39m=\u001b[39m[\u001b[39m'\u001b[39m\u001b[39mgene_id\u001b[39m\u001b[39m'\u001b[39m, file])\n\u001b[0;32m      9\u001b[0m     data[\u001b[39m'\u001b[39m\u001b[39mgene_id\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m data\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m x: x\u001b[39m.\u001b[39mgene_id\u001b[39m.\u001b[39msplit(\u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m'\u001b[39m)[\u001b[39m0\u001b[39m], axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m)\n\u001b[1;32m---> 10\u001b[0m     df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mmerge(df,data, on\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mgene_id\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[0;32m     11\u001b[0m     \u001b[39mprint\u001b[39m(file,\u001b[39m'\u001b[39m\u001b[39m merge completed!\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m     12\u001b[0m \u001b[39mreturn\u001b[39;00m df\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\merge.py:124\u001b[0m, in \u001b[0;36mmerge\u001b[1;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[0;32m     93\u001b[0m \u001b[39m@Substitution\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39mleft : DataFrame or named Series\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     94\u001b[0m \u001b[39m@Appender\u001b[39m(_merge_doc, indents\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m)\n\u001b[0;32m     95\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mmerge\u001b[39m(\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    108\u001b[0m     validate: \u001b[39mstr\u001b[39m \u001b[39m|\u001b[39m \u001b[39mNone\u001b[39;00m \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[0;32m    109\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m DataFrame:\n\u001b[0;32m    110\u001b[0m     op \u001b[39m=\u001b[39m _MergeOperation(\n\u001b[0;32m    111\u001b[0m         left,\n\u001b[0;32m    112\u001b[0m         right,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    122\u001b[0m         validate\u001b[39m=\u001b[39mvalidate,\n\u001b[0;32m    123\u001b[0m     )\n\u001b[1;32m--> 124\u001b[0m     \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mget_result(copy\u001b[39m=\u001b[39;49mcopy)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\merge.py:775\u001b[0m, in \u001b[0;36m_MergeOperation.get_result\u001b[1;34m(self, copy)\u001b[0m\n\u001b[0;32m    771\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mleft, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mright \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_indicator_pre_merge(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mleft, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mright)\n\u001b[0;32m    773\u001b[0m join_index, left_indexer, right_indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_get_join_info()\n\u001b[1;32m--> 775\u001b[0m result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_reindex_and_concat(\n\u001b[0;32m    776\u001b[0m     join_index, left_indexer, right_indexer, copy\u001b[39m=\u001b[39;49mcopy\n\u001b[0;32m    777\u001b[0m )\n\u001b[0;32m    778\u001b[0m result \u001b[39m=\u001b[39m result\u001b[39m.\u001b[39m__finalize__(\u001b[39mself\u001b[39m, method\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_merge_type)\n\u001b[0;32m    780\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mindicator:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\merge.py:766\u001b[0m, in \u001b[0;36m_MergeOperation._reindex_and_concat\u001b[1;34m(self, join_index, left_indexer, right_indexer, copy)\u001b[0m\n\u001b[0;32m    764\u001b[0m left\u001b[39m.\u001b[39mcolumns \u001b[39m=\u001b[39m llabels\n\u001b[0;32m    765\u001b[0m right\u001b[39m.\u001b[39mcolumns \u001b[39m=\u001b[39m rlabels\n\u001b[1;32m--> 766\u001b[0m result \u001b[39m=\u001b[39m concat([left, right], axis\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m, copy\u001b[39m=\u001b[39;49mcopy)\n\u001b[0;32m    767\u001b[0m \u001b[39mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\util\\_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    325\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(args) \u001b[39m>\u001b[39m num_allow_args:\n\u001b[0;32m    326\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[0;32m    327\u001b[0m         msg\u001b[39m.\u001b[39mformat(arguments\u001b[39m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[0;32m    328\u001b[0m         \u001b[39mFutureWarning\u001b[39;00m,\n\u001b[0;32m    329\u001b[0m         stacklevel\u001b[39m=\u001b[39mfind_stack_level(),\n\u001b[0;32m    330\u001b[0m     )\n\u001b[1;32m--> 331\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\concat.py:381\u001b[0m, in \u001b[0;36mconcat\u001b[1;34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[0;32m    159\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    160\u001b[0m \u001b[39mConcatenate pandas objects along a particular axis.\u001b[39;00m\n\u001b[0;32m    161\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    366\u001b[0m \u001b[39m1   3   4\u001b[39;00m\n\u001b[0;32m    367\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    368\u001b[0m op \u001b[39m=\u001b[39m _Concatenator(\n\u001b[0;32m    369\u001b[0m     objs,\n\u001b[0;32m    370\u001b[0m     axis\u001b[39m=\u001b[39maxis,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    378\u001b[0m     sort\u001b[39m=\u001b[39msort,\n\u001b[0;32m    379\u001b[0m )\n\u001b[1;32m--> 381\u001b[0m \u001b[39mreturn\u001b[39;00m op\u001b[39m.\u001b[39;49mget_result()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\reshape\\concat.py:616\u001b[0m, in \u001b[0;36m_Concatenator.get_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    612\u001b[0m             indexers[ax] \u001b[39m=\u001b[39m obj_labels\u001b[39m.\u001b[39mget_indexer(new_labels)\n\u001b[0;32m    614\u001b[0m     mgrs_indexers\u001b[39m.\u001b[39mappend((obj\u001b[39m.\u001b[39m_mgr, indexers))\n\u001b[1;32m--> 616\u001b[0m new_data \u001b[39m=\u001b[39m concatenate_managers(\n\u001b[0;32m    617\u001b[0m     mgrs_indexers, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mnew_axes, concat_axis\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mbm_axis, copy\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcopy\n\u001b[0;32m    618\u001b[0m )\n\u001b[0;32m    619\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcopy:\n\u001b[0;32m    620\u001b[0m     new_data\u001b[39m.\u001b[39m_consolidate_inplace()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python311\\site-packages\\pandas\\core\\internals\\concat.py:212\u001b[0m, in \u001b[0;36mconcatenate_managers\u001b[1;34m(mgrs_indexers, axes, concat_axis, copy)\u001b[0m\n\u001b[0;32m    210\u001b[0m values \u001b[39m=\u001b[39m blk\u001b[39m.\u001b[39mvalues\n\u001b[0;32m    211\u001b[0m \u001b[39mif\u001b[39;00m copy:\n\u001b[1;32m--> 212\u001b[0m     values \u001b[39m=\u001b[39m values\u001b[39m.\u001b[39mcopy()\n\u001b[0;32m    213\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    214\u001b[0m     values \u001b[39m=\u001b[39m values\u001b[39m.\u001b[39mview()\n",
      "\u001b[1;31mMemoryError\u001b[0m: Unable to allocate 353. MiB for an array with shape (2, 23129217) and data type object"
     ]
    }
   ],
   "source": [
    "\n",
    "def all_data(source_dirctory, df):\n",
    "    if not os.path.exists(source_dirctory):\n",
    "        return False\n",
    "    files = os.listdir(source_dirctory)\n",
    "    \n",
    "    for file in files:\n",
    "        \n",
    "        data = pd.read_csv(filepath_or_buffer=source_dirctory + '/' + file,sep=',', names=['gene_id', file])\n",
    "        data['gene_id'] = data.apply(lambda x: x.gene_id.split('.')[0], axis=1)\n",
    "        df = pd.merge(df,data, on='gene_id')\n",
    "        print(file,' merge completed!')\n",
    "    return df\n",
    "\n",
    "\n",
    "dff = all_data(\"D:\\BTp\\Process1data\", GFF3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('mergedgenes1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter only protein coding genes\n",
    "df = pd.read_csv('mergedgenes1.csv')\n",
    "df = df[df['gene_biotype'] == 'protein_coding']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
